lib/stdlib/test/property_test/queue_prop.erl

```diff
@@ -0,0 +1,670 @@
+		    % queue:peek/1, queue:peek_r/1
+		    (Op, {true, Q1, L1}) when Op =:= peek; Op =:= peek_r ->
+			case queue:Op(Q1) of
+			    empty ->
+				{L1 =:= [], Q1, L1};
+			    {value, I} when Op =:= peek ->
+				{I =:= hd(L1), Q1, L1};
+			    {value, I} when Op =:= peek_r ->
+				{I =:= lists:last(L1), Q1, L1}
+			end;
+		    % queue:get/1, queue:head/1
+		    (Op, {true, Q1, L1}) when Op =:= get; Op =:= head ->
+			try queue:Op(Q1) of
+			    I ->
+				{I =:= hd(L1), Q1, L1}
+			catch error:empty ->
+			    {L1 =:= [], Q1, L1}
+			end;
+		    % queue:get_r/1, queue:daeh/1, queue:last/1
+		    (Op, {true, Q1, L1}) when Op =:= get_r; Op =:= daeh; Op =:= last ->
+			try queue:Op(Q1) of
+			    I ->
+				{I =:= lists:last(L1), Q1, L1}
+			catch error:empty ->
+			    {L1 =:= [], Q1, L1}
+			end;
+		    % queue:init/1, queue:drop_r/1, queue:liat/1
+		    (Op, {true, Q1, L1}) when Op =:= init; Op =:= drop_r; Op =:= liat ->
+			try queue:Op(Q1) of
+			    Q2 ->
+				L2 = lists:droplast(L1),
+				{equal(L2, Q2), Q2, L2}
+			catch error:empty ->
+			    {L1 =:= [], Q1, L1}
+			end;
+		    % queue:drop/1, queue:tail/1
+		    (Op, {true, Q1, L1}) when Op =:= drop; Op =:= tail ->
+			try queue:Op(Q1) of
+			    Q2 ->
+				L2 = tl(L1),
+				{equal(L2, Q2), Q2, L2}
+			catch error:empty ->
+			    {L1 =:= [], Q1, L1}
+			end
+		end,
+		{true, Q, L},
+		Ops
+	    ),
+	    Res andalso equal(RL, RQ)
+	end
+    ).
+
+%%%%%%%%%%%%%%%%%%%%%%%%
+%%% Property helpers %%%
+%%%%%%%%%%%%%%%%%%%%%%%%
+
+% get/1 and head/1 have the same semantics
+common_get_head(Fn) ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	try
+	    queue:Fn(Q)
+	of E ->
+	    E =:= hd(L)
+	catch error:empty ->
+	    L =:= []
+	end
+    ).
+
+% get_r/1, last/1 and daeh/1 have the same semantics
+common_get_r_last_daeh(Fn) ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	try
+	    queue:Fn(Q)
+	of E ->
+	    E =:= lists:last(L)
+	catch error:empty ->
+	    L =:= []
+	end
+    ).
+
+% drop_r/1, init/1 and liat/1 have the same semantics 
+common_drop_r_init_liat(Fn) ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	try
+	    queue:Fn(Q)
+	of Q1 ->
+	    equal(lists:droplast(L), Q1)
+	catch error:empty ->
+	    L =:= []
+	end
+    ).
+
+% drop/1 and tail/1 have the same semantics
+common_drop_tail(Fn) ->
+    ?FORALL(
+	{L, Q},
+	list_queue(),
+	try
+	    queue:Fn(Q)
+	of Q1 ->
+	    equal(tl(L), Q1)
+	catch error:empty ->
+	    L =:= []
+	end
+    ).
+
+% in_r/2 and cons/2 have the same semantics
+common_in_r_cons(Fn) ->
+    ?FORALL(
+	L,
+	list(),
+	begin
+	    Q = lists:foldl(
+		fun (I, Acc) ->
+		    queue:Fn(I, Acc)
+		end,
+		queue:new(),
+		L
+	    ),
+	    equal(lists:reverse(L), Q)
+	end
+    ).
+
+%%%%%%%%%%%%%%%%%%
+%%% Generators %%%
+%%%%%%%%%%%%%%%%%%
+
+list_queue() ->
+    list_queue(term()).
+
+list_queue(Type) ->
+    ?LET(List, list(Type), {List, queue:from_list(List)}).
+
+queue() ->
+    queue(term()).
+
+queue(Type) ->
+    ?LET(List, list(Type), queue:from_list(List)).
+
+non_queue() ->
```

Comment:
This generator shows the problem of the `is_queue/1` function, i.e., only by knowing about the internal representation of queues can we write a generator that we can be sure does not generate a term that can be mistaken for a queue.  This is not very abstract and goes against the documentation saying that one should treat the representation as opaque and not rely on a specification representation.

As a side note, this is also true for other modules implementing data structures, e.g., the `sets` and `array` module.

Reply:
Yes... period :sweat: As you point out, it can't be done any different way (or if so, I don't see how). The only alternative I can think of would be to only test positive cases. :thinking: Please advise?

Reply:
There is really no better way to do it, since the problem is with the `is_queue/1` function, for which the result is reliable only when it answers `false`.  I think it is OK to leave it as it is, but add documentation for the enlightenment of future readers, pointing out why the generator is constructed this way.

Reply:
Something we _could_ do is to only rely on the internal representation in `prop_is_queue`, and let the `non_queue` generator rely on `queue:is_queue` for producing non-queues, like `?SUCHTHAT(..., not queue:is_queue(T))`. What do you think?

Reply:
Yes, we _could_ do that, but what are you then testing?  Currently the result of the `non_queue/0` generator is only used to test the `is_queue/1` function.  Using `is_queue/1` in the generator then really doesn't test anything since we are sure of the result.
I'd rather keep it as it is and add a comment about the potential brittleness of the test.

Reply:
Oh, yeah, I forgot that this is the only place where that generator is used :sweat_smile: What In meant was to put the generator as it is in place in `prop_is_queue`, and rewrite the generator to use `queue:is_queue/1` for when it is used in other properties that require non-queues. But since there are no such places, there is no point in doing that.

OTOH, we could add more tests that provide invalid arguments for each function. The `queue` implementation has guards for that on every function, eg https://github.com/erlang/otp/blob/f01ea0ab3d0517fbda145a32346a6cd3310f862a/lib/stdlib/src/queue.erl#L74-L79, something which is currently not covered by the test suite. For that, the `non_queue` generator would be useful.

Reply:
Yes, using a generator to generate terms that are _not_ queues and testing the resilience of the functions is a good idea.  It might this be a good idea to extend the current tests as you suggest.

Reply:
Ok, I added tests for invalid arguments for all functions, take a look :smile: